compiler/damlc/tests/src/DA/Test/ShakeIdeClient.hs

```diff
@@ -409,6 +409,107 @@ dlintSmokeTests mbScenarioService = Tasty.testGroup "Dlint smoke tests"
             setFilesOfInterest [foo]
             expectNoErrors
             expectDiagnostic DsInfo (foo, 2, 11) "Suggestion: Use ::"
+    ,  testCase' "Use guards" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "truth i = if i == 1 then Some True else if i == 2 then Some False else None"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 0) "Suggestion: Use guards"
+    ,  testCase' "Redundant guard" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "foo i | otherwise = True"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 0) "Suggestion: Redundant guard"
+    ,  testCase' "Redundant where" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "foo i = i where"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 0) "Suggestion: Redundant where"
+    ,  testCase' "Use otherwise" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "foo i | i == 1 = True | True = False"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 0) "Suggestion: Use otherwise"
+    ,  testCase' "Use record patterns" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "data Foo = Foo {a : Int, b : Int, c : Int, d : Int}"
+                , "foo (Foo _ _ _ _) = True"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 3, 5) "Suggestion: Use record patterns"
+    ,  testCase' "Used otherwise as a pattern" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "foo otherwise = 1"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 4) "Warning: Used otherwise as a pattern"
+    ,  testCase' "Redundant bang pattern" $ do
```

Comment:
This rule doesn’t make sense for DAML but on the other hand putting bang patterns in DAML never makes sense so I guess we might as well warn about some of them being redundant.

Reply:
Fair. Perhaps more generally `daml-lf` conversion should fail in the presence of bangs? Or maybe we should customize the lexer to fail earlier since they "aren't DAML"?

Reply:
I'll raise an issue for this part of the discussion.